flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/cache/DoubleListLru.java

```diff
@@ -0,0 +1,442 @@
+    private int secondSize;
+    private final HashMap<K, Node> map;
+
+    public DoubleListLru() {
+        this.head = null;
+        this.tail = null;
+        this.middle = null;
+        this.size = 0;
+        this.map = new HashMap<>();
+    }
+
+    // -------------------
+    // Hook methods
+    // -------------------
+
+    /**
+     * Checks if it is safe to add a value to the first list. Will be called before adding a value
+     * to the first list.
+     *
+     * @param value the value to be added to the first list
+     * @return true if it is safe to add the value to the first list, false otherwise.
+     */
+    abstract boolean isSafeToAddFirst(V value);
+
+    /**
+     * Hook method called when a new node is created.
+     *
+     * @param value the value of the new node
+     * @param n the new node
+     */
+    abstract void newNodeCreated(V value, Node n);
+
+    /**
+     * Hook method called when a value is added to the first list.
+     *
+     * @param value the value added to the first list
+     */
+    abstract void addedToFirst(V value);
+
+    /**
+     * Hook method called when a value is added to the second list.
+     *
+     * @param value the value added to the second list
+     */
+    abstract void addedToSecond(V value);
+
+    /**
+     * Hook method called when a value is removed from the first list.
+     *
+     * @param value the value removed from the first list
+     */
+    abstract void removedFromFirst(V value);
+
+    /**
+     * Hook method called when a value is removed from the second list.
+     *
+     * @param value the value removed from the second list
+     */
+    abstract void removedFromSecond(V value);
+
+    /**
+     * Hook method called when a value is moved to the first list.
+     *
+     * @param value the value moved to the first list
+     */
+    abstract void movedToFirst(V value);
+
+    /**
+     * Hook method called when a value is moved to the second list.
+     *
+     * @param value the value moved to the second list
+     */
+    abstract void movedToSecond(V value);
+
+    /**
+     * Hook method called when a node is accessed in the second list.
+     *
+     * @param value the value of the accessed node
+     * @return true if the node should be promoted to the first list, false otherwise
+     */
+    abstract boolean nodeAccessedAtSecond(V value);
+
+    /**
+     * Hook method called when a value is promoted to the first list.
+     *
+     * @param value the promoted value
+     */
+    abstract void promotedToFirst(V value);
+
+    /**
+     * Adds a new entry to the front of the cache.
+     *
+     * @param key the key of the entry
+     * @param value the value of the entry
+     */
+    public void addFirst(K key, V value) {
+        if (!isSafeToAddFirst(value)) {
+            addSecond(key, value);
+            return;
+        }
+        Node newNode = new Node(key, value);
+        newNodeCreated(value, newNode);
+        map.put(key, newNode);
```

Comment:
Can the same key have multiple different values when rescaling down? 
Will the following value overwrite the previous value?

Reply:
No we don't support different value for one key. the cache key is from the source path of MappingEntry, so IIUC in rescaling scenario, we still have one-to-one mapping from key to value here.